server/src/main/java/io/crate/execution/dml/Indexer.java

```diff
@@ -747,38 +801,74 @@ public ParsedDocument index(IndexItem item) throws IOException {
         );
         Object[] values = item.insertValues();
 
-        for (int i = 0; i < values.length; i++) {
-            Reference reference = columns.get(i);
-            Object value = valueForInsert(reference.valueType(), values[i]);
-            ColumnConstraint check = columnConstraints.get(reference.column());
-            if (check != null) {
-                check.verify(value);
-            }
-            if (reference.granularity() == RowGranularity.PARTITION) {
-                continue;
-            }
-            ValueIndexer<Object> valueIndexer = (ValueIndexer<Object>) valueIndexers.get(i);
-            if (value == null) {
-                continue;
+        if (indexOrder.isEmpty()) {
+            for (int i = 0; i < values.length; i++) {
+                Reference reference = columns.get(i);
+                Object value = valueForInsert(reference.valueType(), values[i]);
+                ColumnConstraint check = columnConstraints.get(reference.column());
+                if (check != null) {
+                    check.verify(value);
+                }
+                if (reference.granularity() == RowGranularity.PARTITION) {
+                    continue;
+                }
+                ValueIndexer<Object> valueIndexer = (ValueIndexer<Object>) valueIndexers.get(i);
+                if (value == null) {
+                    continue;
+                }
+                translogWriter.writeFieldName(valueIndexer.storageIdentLeafName());
+                valueIndexer.indexValue(value, docBuilder);
             }
-            translogWriter.writeFieldName(valueIndexer.storageIdentLeafName());
-            valueIndexer.indexValue(value, docBuilder);
-        }
 
-        for (var entry : synthetics.entrySet()) {
-            ColumnIdent column = entry.getKey();
-            if (!column.isRoot()) {
-                continue;
+            for (var entry : synthetics.entrySet()) {
+                ColumnIdent column = entry.getKey();
+                if (!column.isRoot()) {
+                    continue;
+                }
+                Synthetic synthetic = entry.getValue();
+
+                Object value = synthetic.value();
+                if (value == null) {
+                    continue;
+                }
+                ValueIndexer<Object> indexer = synthetic.indexer();
+                translogWriter.writeFieldName(indexer.storageIdentLeafName());
+                indexer.indexValue(value, docBuilder);
             }
-            Synthetic synthetic = entry.getValue();
+        } else {
+            for (Reference ref : indexOrder) {
+                if (columns.contains(ref)) {
+                    int idx = columns.indexOf(ref);
```

Comment:
```suggestion
                int idx = columns.indexOf(ref);
                if (idx >= 0) {
```

Otherwise it does two searches.